native/src/main/java/io/ballerina/stdlib/data/yaml/lexer/CharacterReader.java

```diff
@@ -0,0 +1,178 @@
+     * @return next code point
+     */
+    public int peek() {
+        return peek(0);
+    }
+
+    /**
+     * Peeks the k-th indexed code point.
+     *
+     * @param k number of characters to peek
+     * @return code point at the peek
+     */
+    public int peek(int k) {
+        if (k >= 0 && checkAndReadData(k) && pointer + k >= 0) {
+            return dataBuffer[pointer + k];
+        }
+        return -1;
+    }
+
+    /**
+     * Moves the internal pointer forward by the specified amount (`k`).
+     *
+     * @param k The number of positions to move forward.
+     */
+    public boolean forward(int k) {
+        int i;
+        for (i = 0; i < k && checkAndReadData(k); i++) {
+            int codePoint = dataBuffer[pointer++];
+            if (hasNewLine(codePoint)) {
+                this.remainingBufferedSize -= column + 1;
+                this.column = 0;
+                this.line++;
+            } else if (codePoint != 0xFEFF) {
+                this.column++;
+            }
+        }
+        return i == 0;
+    }
+
+    private boolean hasNewLine(int codePoint) {
+        return codePoint == '\n' || codePoint == '\r' && peek() == '\n';
+    }
+
+    private boolean checkAndReadData(int k) {
+        if (!eof && pointer + k >= dataBufferSize) {
+            readData();
+        }
+        return (pointer + k) < dataBufferSize;
+    }
+
+    private void readData() {
+        try {
+            int size = reader.read(buff);
+            if (size <= 0) {
+                this.eof = true;
+                return;
+            }
+
+            int cpIndex = dataBufferSize - pointer;
+            this.dataBuffer = Arrays.copyOfRange(dataBuffer, pointer, dataBufferSize + size);
+
+            for (int i = 0; i < size; cpIndex++) {
+                int codePoint = Character.codePointAt(buff, i);
+                dataBuffer[cpIndex] = codePoint;
+                if (isPrintable(codePoint)) {
+                    i += Character.charCount(codePoint);
+                } else {
+                    NewLineIndexData newLineIndexData = findLastNewLineIndexAndNewLineCount();
+                    line += newLineIndexData.newLineCount;
+                    column += newLineIndexData.lastNewLineIndex == -1 ? cpIndex
+                            : cpIndex - newLineIndexData.lastNewLineIndex;
+                    throw new Error.YamlParserException("non printable character found", line, column);
+                }
+            }
+            dataBufferSize = cpIndex;
+            remainingBufferedSize = dataBufferSize;
+            pointer = 0;
+        } catch (Error.YamlParserException e) {
+            throw DiagnosticLog.error(DiagnosticErrorCode.YAML_PARSER_EXCEPTION, e.getMessage(), line, column);
+        } catch (IOException e) {
+            throw DiagnosticLog.error(DiagnosticErrorCode.YAML_READER_FAILURE, e.getMessage());
+        }
+    }
+
+    private record NewLineIndexData(int lastNewLineIndex, int newLineCount) {
+    };
```

Comment:
Do we need this? remove if not

Reply:
Yes. We need this. This a record I am using.